src/core/components/pin-set.js

```diff
@@ -0,0 +1,269 @@
+'use strict'
+
+const multihashes = require('multihashes')
+const CID = require('cids')
+const protobuf = require('protons')
+const fnv1a = require('fnv1a')
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const varint = require('varint')
+const once = require('once')
+
+const pbSchema = require('./pin.proto')
+
+const emptyKeyHash = 'QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n'
+const emptyKey = multihashes.fromB58String(emptyKeyHash)
+const defaultFanout = 256
+const maxItems = 8192
+const pb = protobuf(pbSchema)
+
+function toB58String (hash) {
+  return new CID(hash).toBaseEncodedString()
+}
+
+function readHeader (rootNode) {
+  // rootNode.data should be a buffer of the format:
+  // < varint(headerLength) | header | itemData... >
+  const rootData = rootNode.data
+  const hdrLength = varint.decode(rootData)
+  const vBytes = varint.decode.bytes
+  if (vBytes <= 0) {
+    throw new Error('Invalid Set header length')
+  }
+  if (vBytes + hdrLength > rootData.length) {
+    throw new Error('Impossibly large set header length')
+  }
+  const hdrSlice = rootData.slice(vBytes, hdrLength + vBytes)
+  const header = pb.Set.decode(hdrSlice)
+  if (header.version !== 1) {
+    throw new Error(`Unsupported Set version: ${header.version}`)
+  }
+  if (header.fanout > rootNode.links.length) {
+    throw new Error('Impossibly large fanout')
+  }
+  return {
+    header: header,
+    data: rootData.slice(hdrLength + vBytes)
+  }
+}
+
+function hash (seed, key) {
+  const buf = Buffer.alloc(4)
+  buf.writeUInt32LE(seed, 0)
+  const data = Buffer.concat([
+    buf, Buffer.from(toB58String(key))
+  ])
+  return fnv1a(data.toString('binary'))
+}
+
+exports = module.exports = function (dag) {
+  const pinSet = {
+    // should this be part of `object` API?
+    hasChild: (root, childhash, callback, _links, _checked, _seen) => {
+      callback = once(callback)
+      if (typeof childhash === 'object') {
+        childhash = toB58String(childhash)
+      }
+      _links = _links || root.links.length
+      _checked = _checked || 0
+      _seen = _seen || {}
+
+      if (!root.links.length && _links === _checked) {
+        // all nodes have been checked
+        return callback(null, false)
+      }
+      root.links.forEach(link => {
+        const bs58Link = toB58String(link.multihash)
+        if (bs58Link === childhash) {
+          return callback(null, true)
+        }
+
+        // don't check the same links twice
+        if (bs58Link in _seen) { return }
+        _seen[bs58Link] = true
+
+        dag.get(new CID(link.multihash), (err, res) => {
```

Comment:
Do you need to cast it to a CID yourself? I don't think so.

Reply:
It turns out that some of our links' multihashes are buffers and `dag.get` does not know how to handle them. I've added a cast from buffer to CID in core/components/dag.js. Is that why you were thinking it should work without the cast?

https://github.com/ipfs/js-ipfs/pull/1045/files#diff-cb106b8602ebb9750632d0d9ce3c1848R38